### PR TITLE
Credential Fixes Roundup

### PR DESCRIPTION
**Pull request by erikerikson**
## What did you implement:

Merge fixes from both #2373 & #2695, adding editorial suggestions and future-proofing the code a bit (probably)

## How did you implement it:

Merge the great work of @stevecaldwell77 and @rkistner and apply the organizational pattern that had been in place to those changes.

Clean up tests to broadly clean and restore process/test environment and simplify test code.

This work uncovered that we were returning profile values in the profile attribute when no profile credentials were loaded.  This seems inappropriate.  The tests were modified to use the temporary credential file creation code that @stevecaldwell77 wrote exclusively.  This way there are actual credentials to load from the given profile.  Previously it was accepted that we couldn't guarantee an existing profile with our reserved names and values for all contributors.  It seemed this would create a barrier to contribution so it was accepted.  Using the temporary credentials file improves that circumstance although we won't be testing loading from the default location.  I think the responsibility for that working that responsibility rests with the aws-sdk.

## How can we verify it:

I am not sure how to test the assume-role/asynchronous STS credential loading capability and would love some help making sure that is right.  Both of the linked & merged PRs have some instructions on how to verify that capability.

You can set the `AWS_SHARED_CREDENTIALS_FILE` environment variable to test loading credentials from an arbitrary location.  You might rename your `~/.aws/credentials` file and reference the renamed file for example.  Don't forget to rename it back for convenience.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Enable ["Allow edits from maintainers"](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) for this PR
- [x] Change ready for review message below

***Is this ready for review?:*** YES

### CODE REVIEW

**Comment by erikerikson** [comment]
There is one remaining convenience use case that I am aware of.  That is loading profile credentials from the metadata service for cases where serverless is being run on an EC2 instance.  This came up in a CI/CD context for us but we had the tools to solve it easily enough so I haven't a clue about the priority others might assign to the capability.  I believe the SDK would load them itself if we allowed users to proceed without credentials but we don't currently do that.

**Comment by rkistner** [comment]
The `validCredentials` function returns false for my config. It seems like at that point the AWS SDK hasn't even loaded the source_profile from the file yet.

```
SharedIniFileCredentials {
  expired: false,
  expireTime: null,
  accessKeyId: undefined,
  sessionToken: undefined,
  filename: '/home/ralf/.aws/credentials',
  profile: 'myprofile',
  disableAssumeRole: false }
```

It actually seems like `role_arn` and `source_profile` is never actually stored on the credentials object. This is what it looks like after it's loaded (via `credentials.refresh(callback)`):

```
SharedIniFileCredentials {
  expired: false,
  expireTime: Tue Nov 29 2016 14:48:02 GMT+0200 (SAST),
  accessKeyId: '<key id here>',
  sessionToken: '<token here>',
  filename: '/home/ralf/.aws/credentials',
  profile: 'myprofile',
  disableAssumeRole: false }
```

The only feasible check seems to be for `credentials.profile`.

**Comment by erikerikson** [comment]
Crap.  That's what I was seeing.  I'm sorry.  I was hoping that was a result of my set up.  The case I am concerned about with checking the profile attribute is where the profile doesn't exist in the target credentials file so we get funky results.  That can create a lot of confusion.

Do you explicitly call refresh usually or let the sdk manage that for you?

**Comment by rkistner** [comment]
The easiest is to let the SDK manage it for you, but then you only see the results on the first SDK call. On the other hand, if you call `refresh` explicitly, you need to wait for the asynchronous result.

**Comment by erikerikson** [comment]
Thanks for confirming.  I've got to get some sleep but I will make some time to get this sorted tomorrow. Just changing the second condition statement of validCredentials might do the trick but I'd like to bang on it a bit before committing to that approach

**Comment by erikerikson** [comment]
Regarding (added `secretAccessKey`):

> SharedIniFileCredentials {
  expired: false,
  expireTime: null,
  accessKeyId: undefined,
  [secretAccessKey: undefined] // this is unenumerable but there
  sessionToken: undefined,
  filename: '/home/ralf/.aws/credentials',
  profile: 'myprofile',
  disableAssumeRole: false }

So I gave it an non-existant profile and ended up with the following:
```
{
  "expired": false,
  "expireTime": null,
  "filename": "/var/folders/_r/kr98c8kd4wb6zwsz5j3zfl0xb2b151/T/tmpdirs-serverless/serverless/ba0457d8cf8e9b0d/credentials",
  "profile": "notDefaultAsyncFOO",
  "disableAssumeRole": false
  "accessKeyId": undefined,
  "secretAccessKey": undefined
  "sessionToken": undefined,
}
```

So it seems that the case of "there's no profile there" is indistinguishable from the "hold on, I'm getting it" case.

**Comment by erikerikson** [comment]
Issue for indistinguishability created in aws-sdk: https://github.com/aws/aws-sdk-js/issues/1243
Pondering whether a short term work around or a quick PR to aws-sdk-js would be better

**Comment by stevecaldwell77** [comment]
Selfishly I vote for a short-term workaround, because I would bet on it being released earlier than a PR to aws-sdk-js :)

I need this functionality to make serverless work for me (our security model is dependent upon assuming roles), so I currently have to use a forked version where I've applied @rkistner's changes.

**Comment by erikerikson** [comment]
:D  I don't disagree @stevecaldwell77.  I'd appreciate some sign off from @flomotlik, @pmuens, or company.

The sign off I am talking about would be choosing to accept loading invalid credentials from profiles which were explicitly specified but do not exist.  Specifying a profile that doesn't exist would result in the loading and attempted use of empty/null credentials.  This would be a temporary state of things while we wait for [the fix on aws-sdk-js](https://github.com/aws/aws-sdk-js/pull/1244) to go through.  After that point, we will be able to actively detect the initial validation of asynch credentials and reject any invalid profile specifications.  Perhaps someone with more knowledge or creativity than I (trying to do too many things at once) am coming up with.

**Comment by chrisradek** [comment]
aws/aws-sdk-js#1243 has just been merged if that changes your plans :)

**Comment by stevecaldwell77** [comment]
@chrisradek - 👍  👍   I'm happy to lose the theoretical bet in my above comment :)

**Comment by erikerikson** [comment]
I will get this in tonight, pub crawl or not!  :D

**Comment by erikerikson** [comment]
Thanks @chrisradek!

**Comment by erikerikson** [comment]
Thank you to everyone that has been involved!  🎉 

This should not yet be merged although it is ready.  The reason is that it depends on a yet-to-be published version of `aws-sdk`.  Currently the version is `v2.7.9` and the package.json has been changed to demand `^v2.7.10`.  Once the latest version of aws-sdk-js master is published, this should be GTM.

To test this prior to publishing of `^v2.7.10`, run:
`npm install git+https://git@github.com/aws/aws-sdk-js.git --save`

Then running the standard `npm run test` will run a test that confirms the positive and negative functionality via the tests in commit 32100aff.

**Comment by erikerikson** [comment]
I `push --force`'d a missing semi-colon (lint error) and it appears to have stopped up travis.  Someone may need to nudge it.

**Comment by pmuens** [comment]
Thanks for working on this @erikerikson! Just restarted the build!

**Comment by erikerikson** [comment]
Note to self (and anyone observing): this is failing because the future version of aws-sdk is not yet published.  Therefore the code pulled into the travis container doesn't have the roleArn defining code from aws-sdk.  Carry on, keep monitoring.

**Comment by pmuens** [comment]
Thanks for keeping us posted @erikerikson 👍 

Will review once the new `aws-sdk` version is available (please write a note here so that we can start to look at it ASAP)!

**Comment by erikerikson** [comment]
That merge commit addresses the conflict.  Coincidentally, it also pulls in the released fixes in the aws-sdk.

This PR is ready to merge.

**Comment by pmuens** [comment]
@erikerikson thanks for the update! 👍 

The PR will be in our review pipeline for an upcoming release!

**Comment by rowanu** [comment]
Nice work on this one @erikerikson 😃 

Definitely want this in 1.5 please!

File: docs/providers/aws/guide/credentials.md [hunk 1]
```diff
@@ -95,7 +95,7 @@ Default region name [None]: us-west-2
 Default output format [None]: ENTER
 ```
 
-Credentials are stored in INI format in `~/.aws/credentials`, which you can edit directly if needed. Read more about that file in the [AWS documentation](http://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html#cli-config-files)
+Credentials are stored in INI format in `~/.aws/credentials`, which you can edit directly if needed. You can change the path to the credentials file via the AWS_SHARED_CREDENTIALS_FILE environment variable. Read more about that file in the [AWS documentation](http://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html#cli-config-files)
```
Comment (pmuens) [inline docs/providers/aws/guide/credentials.md]: Maybe we should add a comment that the `profile` property needs to be set (e.g. in the `serverless.yml` file) when this environment variable is set.

**Comment by pmuens** [comment]
Just one comment while looking through the code. I can remember a cicrumstance where our keys were exposed in Travis when tests regarding the `getCredentials()` call failed (See https://github.com/serverless/serverless/pull/2294 and https://github.com/serverless/serverless/issues/2308). I have not looked deeply into this right now. But we need to ensure that this is still not the case when this PR is merged...

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 2]
```diff
@@ -191,46 +192,95 @@ describe('AwsProvider', () => {
...
+      AWS_ACCESS_KEY_ID: 'undefined',
+      AWS_SECRET_ACCESS_KEY: 'undefined',
+      AWS_SESSION_TOKEN: 'undefined',
+      AWS_TESTSTAGE_ACCESS_KEY_ID: 'undefined',
+      AWS_TESTSTAGE_SECRET_ACCESS_KEY: 'undefined',
+      AWS_TESTSTAGE_SESSION_TOKEN: 'undefined',
+      AWS_SHARED_CREDENTIALS_FILE: testUtils.getTmpFilePath('credentials'),
+      AWS_PROFILE: 'undefined',
+      AWS_TESTSTAGE_PROFILE: 'undefined',
+    };
```
Comment (eahefnawy) [inline lib/plugins/aws/provider/awsProvider.test.js]: Love it! ❤️

**Comment by eahefnawy** [comment]
@pmuens also about #2308, even though we're still mocking, I'd say that this makes the tests much more robust despite this mocking. The issue is that we'll always need to mock to stay secure otherwise we'd be testing real keys.

So yeah, I'd say this pretty much resolves it.

**Pull request merged**